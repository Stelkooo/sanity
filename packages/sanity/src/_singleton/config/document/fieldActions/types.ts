import {type I18nTextRecord} from '@sanity/types'
import {type ComponentType, type ReactNode} from 'react'

/**
 * @hidden
 * @beta */
export type DocumentFieldActionStatus = 'info' | 'success' | 'warning' | 'error'

/**
 * @hidden
 * @beta */
export type DocumentFieldActionTone = 'primary' | 'positive' | 'caution' | 'critical'

/**
 * @hidden
 * @beta */
export interface DocumentFieldActionItem {
  type: 'action'
  disabled?: boolean | {reason: ReactNode}
  hidden?: boolean
  icon?: ComponentType
  iconRight?: ComponentType
  onAction: () => void
  renderAsButton?: boolean
  selected?: boolean
  status?: DocumentFieldActionStatus
  title: string
  i18n?: I18nTextRecord<'title'>
  tone?: DocumentFieldActionTone
}

/**
 * @hidden
 * @beta */
export interface DocumentFieldActionGroup {
  type: 'group'
  children: DocumentFieldActionNode[]
  disabled?: boolean | {reason: ReactNode}
  expanded?: boolean
  hidden?: boolean
  icon?: ComponentType
  renderAsButton?: boolean
  status?: DocumentFieldActionStatus
  title: string
  i18n?: I18nTextRecord<'title'>
  tone?: DocumentFieldActionTone
}

/**
 * @hidden
 * @beta */
export interface DocumentFieldActionDivider {
  type: 'divider'
}

/**
 * @hidden
 * @beta */
export type DocumentFieldActionNode =
  | DocumentFieldActionItem
  | DocumentFieldActionGroup
  | DocumentFieldActionDivider
