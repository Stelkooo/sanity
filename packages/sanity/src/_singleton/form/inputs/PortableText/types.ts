import {type MutableRefObject, type ReactNode} from 'react'

import {type ArrayOfObjectsItemMember} from '../../store/types/members'
import {type ObjectFormNode} from '../../store/types/nodes'

/** @internal */
export type PortableTextEditorElement = HTMLDivElement | HTMLSpanElement

/** @internal */
export interface PortableTextMemberItem {
  kind: 'annotation' | 'textBlock' | 'objectBlock' | 'inlineObject'
  key: string
  member: ArrayOfObjectsItemMember
  node: ObjectFormNode
  elementRef?: MutableRefObject<PortableTextEditorElement | null>
  input?: ReactNode
}
