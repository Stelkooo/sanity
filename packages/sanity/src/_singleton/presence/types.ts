import {type Path, type User} from '@sanity/types'

/**
 * @hidden
 * @public */
export interface FormNodePresence {
  user: User
  path: Path
  sessionId: string
  lastActiveAt: string
}
