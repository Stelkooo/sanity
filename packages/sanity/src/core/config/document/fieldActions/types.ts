import {type Path, type SchemaType} from '@sanity/types'
import {type DocumentFieldActionGroup, type DocumentFieldActionItem} from 'sanity/_singleton'

import {type ComposableOption, type ConfigContext} from '../../types'

/**
 * @hidden
 * @beta */
export interface DocumentFieldAction {
  name: string
  useAction: DocumentFieldActionHook
}

/**
 * @hidden
 * @beta */
export interface DocumentFieldActionHook {
  (props: DocumentFieldActionProps): DocumentFieldActionItem | DocumentFieldActionGroup
}

/**
 * @hidden
 * @beta */
export interface DocumentFieldActionProps {
  documentId: string
  documentType: string
  path: Path
  schemaType: SchemaType
}

/**
 * @hidden
 * @beta */
export interface DocumentFieldActionsResolverContext extends ConfigContext {
  documentId: string
  documentType: string
  schemaType: SchemaType
}

/**
 * @hidden
 * @beta */
export type DocumentFieldActionsResolver = ComposableOption<
  DocumentFieldAction[],
  DocumentFieldActionsResolverContext
>
