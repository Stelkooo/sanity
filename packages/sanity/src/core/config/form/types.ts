/**
 * @hidden
 * @beta */

import {ComponentType} from 'react'
import {PreviewProps} from '../../components'
import {InputProps, FieldProps, ItemProps, BlockProps, BlockAnnotationProps} from '../../form'

/** @beta */
export interface FormLayoutProps {
  documentId: string
  documentType: string
  renderDefault: (props: FormLayoutProps) => React.ReactElement
}

/**
 * @hidden
 * @beta */
export interface FormComponents {
  annotation?: ComponentType<BlockAnnotationProps>
  block?: ComponentType<BlockProps>
  field?: ComponentType<FieldProps>
  inlineBlock?: ComponentType<BlockProps>
  input?: ComponentType<InputProps>
  item?: ComponentType<ItemProps>
  layout?: ComponentType<FormLayoutProps>
  preview?: ComponentType<PreviewProps>
}
