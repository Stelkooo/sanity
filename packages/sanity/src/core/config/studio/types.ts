import type {InitOptions, ResourceLanguage} from 'i18next'
import React from 'react'
import {ConfigContext, LanguageDefinition, LanguageLoader, Tool} from '../types'

/**
 * @hidden
 * @beta */
// Components
export interface LayoutProps {
  renderDefault: (props: LayoutProps) => React.ReactElement
}

/**
 * @hidden
 * @beta */
export interface LogoProps {
  title: string
  renderDefault: (props: LogoProps) => React.ReactElement
}

/**
 * @hidden
 * @beta */
export interface NavbarProps {
  renderDefault: (props: NavbarProps) => React.ReactElement
}

/**
 * @hidden
 * @beta */
export interface ToolMenuProps {
  activeToolName?: string
  closeSidebar: () => void
  context: 'sidebar' | 'topbar'
  isSidebarOpen: boolean
  tools: Tool[]
  renderDefault: (props: ToolMenuProps) => React.ReactElement
}

/**
 * @hidden
 * @beta */
// Config
export interface StudioComponents {
  layout: React.ComponentType<Omit<LayoutProps, 'renderDefault'>>
  logo: React.ComponentType<Omit<LogoProps, 'renderDefault'>>
  navbar: React.ComponentType<Omit<NavbarProps, 'renderDefault'>>
  toolMenu: React.ComponentType<Omit<ToolMenuProps, 'renderDefault'>>
}

/**
 * @hidden
 * @beta */
export interface StudioComponentsPluginOptions {
  layout?: React.ComponentType<LayoutProps>
  logo?: React.ComponentType<LogoProps>
  navbar?: React.ComponentType<NavbarProps>
  toolMenu?: React.ComponentType<ToolMenuProps>
}

/** @alpha */
export interface I18nContext {
  projectId: string
  dataset: string
}

/** @alpha */
export interface LanguageBundle {
  namespace: string
  resources: ResourceLanguage
  /** Should the resources be merged deeply (nested objects). Default: true */
  deep?: boolean
  /** Should existing resource keys for the namespace be overwritten. Default: false */
  overwrite?: boolean
}

/** @alpha */
export type I18nLanguagesOption =
  | ((prev: LanguageDefinition[], context: I18nContext) => LanguageDefinition[])
  | LanguageDefinition[]

/** @alpha */
export type I18nLanguageLoaderOption =
  | ((prev: LanguageLoader[], context: I18nContext) => LanguageLoader[])
  | LanguageLoader[]

/** @alpha */
export interface I18nPluginOptions {
  /**
   * Defines which languages should be available for user selection.
   * Prev is initially `[{id: 'en-US', title: 'English (US)', icon: AmericanFlag }]`
   *
   * Language titles and icons can be changed by transforming the LanguageDefinition array values.
   *
   * User selected language
   */
  languages?: I18nLanguagesOption

  /**
   * Allows redefining the I18next init options before they are used.
   * Invoked when a workspace is loaded
   */
  initOptions?: (options: InitOptions, context: I18nContext) => InitOptions

  /**
   * Defines language bundles that will be loaded lazily.
   *
   * ### Example
   *
   * ```ts
   *
   * ```
   *
   */
  languageLoaders?: I18nLanguageLoaderOption

  /**
   * When this is true, schema type title and descriptions will be translated.
   * Configure a languageLoader that returns a language bundle for the `schema` namespace,
   * with resource keys using the following convention:
   *
   * ## Keys for types
   * - `<typeName>|title`
   * - `<typeName>|description`
   *
   * ## Keys for fields
   *
   * -`<objectTypeName>.<fieldName>|title`
   * -`<objectTypeName>.<fieldName>|description`
   *
   * ## Keys for array items
   *
   * - `<arrayTypeName>.<arrayMemberTypeName>|title`
   * - `<arrayTypeName>.<arrayMemberTypeName>|description`
   *
   * ## Caveats
   *
   * Enabling schema translations could adversely impact studio performance.
   * Inline definitions for objects are not supported (nested types).
   *
   * ## Example LanguageBundle
   *
   *```ts
   * // locales/en_US/schema.ts
   * export default {
   *   namespace: 'schema',
   *   resources: {
   *     'myDocumentType|title': 'Document type 'myDocumentType' will use this string as title wherever it is used',
   *
   *     'myDocumentType.text|title': 'Document field named 'text' will use this string as title'
   *     'myDocumentType.text|description': 'Document field named 'text' will this string as description',
   *   },
   *  }
   *```
   *
   */
  experimentalTranslateSchemas?: boolean
}
