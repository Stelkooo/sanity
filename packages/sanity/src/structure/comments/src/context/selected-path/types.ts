export interface CommentsSelectedPath {
  origin: 'form' | 'inspector' | 'url'
  fieldPath: string | null
  threadId: string | null
  target?: 'field' | 'block'
}

export interface CommentsSelectedPathContextValue {
  setSelectedPath: (nextSelectedPath: CommentsSelectedPath | null) => void
  selectedPath: CommentsSelectedPath | null
}
