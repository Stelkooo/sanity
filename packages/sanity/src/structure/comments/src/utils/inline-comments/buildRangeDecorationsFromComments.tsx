/* eslint-disable max-depth */
import {RangeDecoration, EditorSelection} from '@sanity/portable-text-editor'
import {isPortableTextTextBlock, isPortableTextSpan, PortableTextTextBlock} from '@sanity/types'
import {fuzzy} from 'fast-fuzzy'
import {flatten} from 'lodash'
import {PropsWithChildren} from 'react'
import {CommentMessage, CommentThreadItem} from '../../types'

// const CHILD_SYMBOL = '\uF0D0'
const CHILD_SYMBOL = '|'
const EMPTY_ARRAY: [] = []
const FALLBACK_TO_BLOCK = true

function CommentDecorator(props: PropsWithChildren<{commentId: string}>) {
  const {children, commentId} = props
  return (
    <span
      data-inline-comment-state="added"
      data-inline-comment-id={commentId}
      style={{backgroundColor: '#ffcc00', color: '#000'}}
    >
      {children}
    </span>
  )
}

function toPlainTextWithChildSeparators(inputBlock: PortableTextTextBlock) {
  return inputBlock.children.map((child) => child.text).join(CHILD_SYMBOL)
}

export interface BuildCommentsRangeDecorationsProps {
  value: CommentMessage
  comments: CommentThreadItem[]
}

/**
 * A function that builds range decorations from comments and the comment message value
 */
export function buildRangeDecorationsFromComments(
  props: BuildCommentsRangeDecorationsProps,
): RangeDecoration[] {
  const {value, comments} = props

  if (!value || value.length === 0) return EMPTY_ARRAY

  const textSelections = comments.filter(
    (comment) => comment.selection?.type === 'text' && comment.selection.value,
  )

  const decorators = textSelections.map((comment) => {
    const decoratorRanges = comment.selection?.value.map((range) => {
      const matchedBlock = value.find((block) => block._key === range._key)

      if (!matchedBlock || !isPortableTextTextBlock(matchedBlock)) {
        return EMPTY_ARRAY
      }

      const positions: EditorSelection[] = []
      const isMatched = false

      if (typeof range.text === 'string') {
        const text = toPlainTextWithChildSeparators(matchedBlock)
        const matchData = fuzzy(range.text, text, {returnMatchData: true})
        const matchPosition = matchData.match.index
        // console.log(JSON.stringify(matchData, null, 2))

        if (matchPosition > -1 && matchData.score > 0.8) {
          let childIndexAnchor = 0
          let anchorOffset = 0
          let childIndexFocus = 0
          let focusOffset = 0
          for (let i = 0; i < matchData.original.length; i++) {
            if (matchData.original[i] === CHILD_SYMBOL) {
              if (i <= matchData.match.index) {
                anchorOffset = -1
                childIndexAnchor++
              }
              childIndexFocus++
              focusOffset = -1
            }
            if (i < matchData.match.index) {
              anchorOffset++
            }
            if (i < matchData.match.index + matchData.match.length) {
              focusOffset++
            }
            if (i === matchData.match.index + matchData.match.length) {
              break
            }
          }

          const decorationSelection: RangeDecoration['selection'] = {
            anchor: {
              path: [
                {_key: matchedBlock._key},
                'children',
                {_key: matchedBlock.children[childIndexAnchor]._key},
              ],
              offset: anchorOffset,
            },
            focus: {
              path: [
                {_key: matchedBlock._key},
                'children',
                {_key: matchedBlock.children[childIndexFocus]._key},
              ],
              offset: focusOffset,
            },
          }

          positions.push(decorationSelection)

          return positions
        }
      }

      if ((!isMatched && FALLBACK_TO_BLOCK) || !range.text) {
        let endOffset = 0
        const lastChild = matchedBlock.children[matchedBlock.children.length - 1]

        if (isPortableTextSpan(lastChild)) {
          endOffset = lastChild.text.length
        }

        const decorationSelection: RangeDecoration['selection'] = {
          anchor: {
            path: [{_key: matchedBlock._key}, 'children', {_key: matchedBlock.children[0]._key}],
            offset: 0,
          },
          focus: {
            path: [
              {_key: matchedBlock._key},
              'children',
              {_key: matchedBlock.children[matchedBlock.children.length - 1]._key},
            ],
            offset: endOffset,
          },
        }

        positions.push(decorationSelection)

        return positions
      }

      return EMPTY_ARRAY
    })

    return flatten(decoratorRanges).map((range) => {
      const rangeDecoration: RangeDecoration = {
        component: ({children}) => (
          <CommentDecorator commentId={comment.parentComment._id}>{children}</CommentDecorator>
        ),
        isRangeInvalid: () => false,
        selection: range,
      }

      return rangeDecoration
    })
  })

  return flatten(decorators)
}
