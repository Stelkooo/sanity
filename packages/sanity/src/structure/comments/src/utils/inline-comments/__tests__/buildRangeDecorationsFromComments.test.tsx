import {renderHook} from '@testing-library/react'
import {buildRangeDecorationsFromComments} from '../buildRangeDecorationsFromComments'
import {CommentThreadItem} from '../../../types'

describe('buildRangeDecorationsFromComments', () => {
  it('can be tested', () => {
    const {result} = renderHook(() => buildRangeDecorationsFromComments({value: [], comments: []}))
    expect(result.current).toEqual([])
  })
  it('creates initial ranges', () => {
    const {result} = renderHook(() =>
      buildRangeDecorationsFromComments({value: initialValue, comments: initialComments}),
    )
    expect(result.current).toMatchSnapshot()
  })
  it('allows something to be bolded before the range', () => {
    const {result} = renderHook(() =>
      buildRangeDecorationsFromComments({value: boldedBeforeValue, comments: initialComments}),
    )
    expect(result.current).toMatchSnapshot()
  })
  it('allows something to be bolded inside the range', () => {
    const {result} = renderHook(() =>
      buildRangeDecorationsFromComments({value: boldedInsideValue, comments: initialComments}),
    )
    expect(result.current).toMatchSnapshot()
  })
})

const initialValue = [
  {
    _key: '6222e4072b6e',
    children: [
      {
        _type: 'span',
        marks: [],
        text: 'Hello there world',
        _key: '9d9c95878a6e0',
      },
    ],
    markDefs: [],
    _type: 'block',
    style: 'normal',
  },
]

const boldedBeforeValue = [
  {
    _key: '6222e4072b6e',
    children: [
      {
        _type: 'span',
        marks: ['strong'],
        text: 'Hello',
        _key: '9d9c95878a6e0',
      },
      {
        _type: 'span',
        marks: [],
        _key: '5d176cf77466',
        text: ' there world',
      },
    ],
    markDefs: [],
    _type: 'block',
    style: 'normal',
  },
]

const boldedInsideValue = [
  {
    _key: '6222e4072b6e',
    children: [
      {
        _type: 'span',
        marks: [],
        text: 'Hello th',
        _key: '9d9c95878a6e0',
      },
      {
        _type: 'span',
        marks: ['strong'],
        _key: 'ea97036ed5c4',
        text: 'e',
      },
      {
        _type: 'span',
        marks: [],
        _key: '8daa33e86194',
        text: 're world',
      },
    ],
    markDefs: [],
    _type: 'block',
    style: 'normal',
  },
]

const initialComments: CommentThreadItem[] = [
  {
    selection: {
      type: 'text',
      value: [
        {
          _key: '6222e4072b6e',
          text: 'there',
        },
      ],
    },
    breadcrumbs: [],
    commentsCount: 0,
    fieldPath: '',
    replies: [],
    parentComment: {
      _id: 'a14fb200-a4c4-4c44-b2c1-c17aa6d79aa8',
      _type: 'comment',
      _createdAt: '',
      _rev: '',
      authorId: '',
      message: null,
      threadId: '',
      status: 'open',
      reactions: null,
      target: {
        path: {
          field: '',
        },
        documentType: '',
        document: {
          _dataset: '',
          _projectId: '',
          _ref: '',
          _type: 'crossDatasetReference',
          _weak: false,
        },
      },
    },
    threadId: 'de1f947c-93a7-4f71-9b9c-5f20023f9c98',
  },
]
