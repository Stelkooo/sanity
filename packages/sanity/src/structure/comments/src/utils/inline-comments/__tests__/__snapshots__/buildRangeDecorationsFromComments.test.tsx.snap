// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildRangeDecorationsFromComments allows something to be bolded before the range 1`] = `
Array [
  Object {
    "anchor": Object {
      "offset": 1,
      "path": Array [
        Object {
          "_key": "6222e4072b6e",
        },
        "children",
        Object {
          "_key": "5d176cf77466",
        },
      ],
    },
    "focus": Object {
      "offset": 6,
      "path": Array [
        Object {
          "_key": "6222e4072b6e",
        },
        "children",
        Object {
          "_key": "5d176cf77466",
        },
      ],
    },
  },
]
`;

exports[`buildRangeDecorationsFromComments allows something to be bolded inside and outside of the range 1`] = `
Array [
  Object {
    "anchor": Object {
      "offset": 3,
      "path": Array [
        Object {
          "_key": "6222e4072b6e",
        },
        "children",
        Object {
          "_key": "897d8881c889",
        },
      ],
    },
    "focus": Object {
      "offset": 3,
      "path": Array [
        Object {
          "_key": "6222e4072b6e",
        },
        "children",
        Object {
          "_key": "3b404dd88fc1",
        },
      ],
    },
  },
]
`;

exports[`buildRangeDecorationsFromComments allows something to be bolded inside the range 1`] = `
Array [
  Object {
    "anchor": Object {
      "offset": 6,
      "path": Array [
        Object {
          "_key": "6222e4072b6e",
        },
        "children",
        Object {
          "_key": "9d9c95878a6e0",
        },
      ],
    },
    "focus": Object {
      "offset": 2,
      "path": Array [
        Object {
          "_key": "6222e4072b6e",
        },
        "children",
        Object {
          "_key": "8daa33e86194",
        },
      ],
    },
  },
]
`;

exports[`buildRangeDecorationsFromComments creates initial ranges 1`] = `
Array [
  Object {
    "anchor": Object {
      "offset": 6,
      "path": Array [
        Object {
          "_key": "6222e4072b6e",
        },
        "children",
        Object {
          "_key": "9d9c95878a6e0",
        },
      ],
    },
    "focus": Object {
      "offset": 11,
      "path": Array [
        Object {
          "_key": "6222e4072b6e",
        },
        "children",
        Object {
          "_key": "9d9c95878a6e0",
        },
      ],
    },
  },
]
`;
